### DISCUSSION TITLE
Future Rule 2b computed size change?

### DISCUSSION THREAD

**Discussion by SirWumpus** [ioccc-src/mkiocccentry#903] [General]
I've been thinking about this the past couple of years and how the judges feel about the following question: `Does compacting a program to use single letters for identifiers really help with obfuscation any more or is it a tired "trope"?`

Back in the day compacting functions and variables to single letters helped get the program within the size limits of Rule 2.  There was a time too that this would aid in obfuscation (I remember one winner using combinations of `o` for all the identifiers).  But has its time as an obfuscation technique become *tired*?  

Do the judges still find this interesting or are they now more interested in interesting concepts, data structures, and techniques and less so in the visual appearance (excluding the best layout winners).  Would the use of proper meaningful (for some definition thereof) identifiers be just as good?

Imagine for example someone using all the words from "Jabberwocky" as identifiers in some order.  You'd blow Rule 2 limit easy, but it should prove different.  Is it worth considering for some future contest (not necessarily the next, maybe the next next one?) adding a new clause to Rule 2b such that:

`All identifiers count at most two (2) octets towards Rule 2b.?`

Not sure if C library et al. should be count as-is or subject to the rule, just the developer's identifiers?

This would mean programs would start to appear more de-obfuscated for reading, while still bound in size.  With a refocus on teaching and learning is this something to consider?  (Maybe a new side contest? The IDCCC = Intl. Deobfuscated C Code Contest 😁).

**Comment by xexyl** [comment]
Yes it does especially if it's something that looks like something else. `l` can, depending on font, look like the digit 1. I actually have done this and I'm sure others have too.

That's just one example. Also it makes the program more compact even if it's not more obscure. That has some nice benefits too.

Plus the size increase is not at this point final as they have said.

Besides, even if it doesn't help obscure the code it doesn't mean it's bad to have it. I think it does, though, because it's a lot harder to keep track of single letter identifiers than names.

Also, as they say, they like well rounded obfuscation. If it was just single letter identifiers that would not win anyway.

**Comment by SirWumpus** [comment]
Another example would be an entry where the identifiers are UTF-8 multibyte characters, consider how obfuscated an entry would look written in mix of Japanese, German, and Sanskrit.  It certainly would be obfuscated to most readers.

**Reply by xexyl** [reply]
> Another example would be an entry where the identifiers are UTF-8 multibyte characters, consider how obfuscated an entry would look written in mix of Japanese, German, and Sanskrit. It certainly would be obfuscated to most.

The 'private tool' I was asked to write (and that's all I'll say :-) ) actually has some German in it but not noticeable as German.

There certainly are cases where longer names can help with obfuscation but so too can single letters. It's not about JUST identifier names either.

Consider this too: it's been over four years since the contest was open. I bet many people have already written code for the contest and to make them not use single letter identifiers would be really cruel and not fair.

**Comment by SirWumpus** [comment]
> Also it makes the program more compact even if it's not more obscure. That has some nice benefits too.

Compact code is not necessarily interesting any more, consider the Rule 2b changes to move away from "blob" code.

**Reply by xexyl** [reply]
> > Also it makes the program more compact even if it's not more obscure. That has some nice benefits too.
> 
> 
> 
> Compact code is not necessarily interesting any more, consider the Rule 2b changes to move away from "blob" code. 

True but the judges also don't like code that is bigger than it has to be. 

Compact code doesn't have to be a blob either. It certainly can be but just because code is compact doesn't mean it's blobby. 

I do like code in the contest that has longer identifiers but I also appreciate single character names. In the case of my entries it helps fit them on shirts. But that part is obviously a more personal reason (that also doesn't change anything above).

**Comment by SirWumpus** [comment]
> to make them not use single letter identifiers would be really cruel and not fair.

I'm not saying to disallow single letters, just allow moving away from it.  Note:

`All identifiers count at most two (2) octets towards Rule 2b.?`

I said `... at most two ...` so if you used single letters, they would count as one still.

**Reply by xexyl** [reply]
> > to make them not use single letter identifiers would be really cruel and not fair.
> 
> I'm not saying to disallow single letters, just allow moving away from it. Note:
> 
> `All identifiers count at most two (2) octets towards Rule 2b.?`
> 
> I said `... at most two ...` so if you used single letters, they would count as one still.

That's fair enough .. I am working on some technical details so I was only briefly looking.  But of course one could already move away from it - there are many entries that do not use single letter identifiers even.

**Reply by xexyl** [reply]
> > > to make them not use single letter identifiers would be really cruel and not fair.
> > 
> > 
> > I'm not saying to disallow single letters, just allow moving away from it. Note:
> > `All identifiers count at most two (2) octets towards Rule 2b.?`
> > I said `... at most two ...` so if you used single letters, they would count as one still.
> 
> That's fair enough .. I am working on some technical details so I was only briefly looking. But of course one could already move away from it - there are many entries that do not use single letter identifiers even.

Technical **fixes** I should actually say, and I'm trying to finish them up before I leave (soon) as I have something else to do before that too.

**Reply by SirWumpus** [reply]
Or maybe the change to make is to count characters instead of octets.  So multiple byte characters would count as one (not 1, 2, 3, or 4 octets).  Japanese has 96 (?) unaccented hiragana glyphs as I recall.  That still would make for interesting code to read.

**Reply by SirWumpus** [reply]
Sometimes I wonder if I have just too warped a sense of humour (or torture).

**Reply by xexyl** [reply]
> Sometimes I wonder if I have just too warped a sense of humour (or torture).

You're not alone. And I say there is no such thing as too much of a warped sense of humour. Unless of course you're doing it to deliberately hurt people. But that's when it's not funny. It doesn't mean you can't have the same sense of humour - you just have to be careful to not say/do/whatever everywhere (and certainly not publicly).

**Reply by SirWumpus** [reply]
To restate a point I made above that might having been overlooked:

> Or maybe the change to make is to count characters instead of octets. 

The above minor change would be the simplest evolution of Rule 2 without drastically changing the rules as proposed by OP.  So ...
- - -
Reading a C source file from standard input, apply the IOCCC source size rules as explained in the Guidelines.  The source code is passed through on standard output.  The source's net ***character*** length is written to standard error; with -v option the net ***character*** length, gross ***character*** length, and matched keyword counts are written.

The entry's gross size in ~~bytes~~ ***characters*** must be less than equal to 4993 ~~bytes~~ ***characters*** in length.

The entry's net size in ~~bytes~~ ***characters*** must be less than equal to 2503 ~~bytes~~ ***characters***.  The net size is computed as follows:

The size tool counts most C reserved words (keyword, secondary, and selected preprocessor keywords) as 1.  The size tool counts all other ~~octets~~ ***characters*** as 1 excluding ASCII whitespace, and excluding any ‘;’, ‘{’ or ‘}’ followed by ASCII whitespace, and excluding any ‘;’, ‘{’ or ‘}’ ~~octet~~ ***character*** immediately before the end of file.
- - -
The should allow for better UTF8 multibyte support, which the C Standard has supported for some time now.  This means that non-english speakers could present entries where function and variable names are no longer ASCII characters, but still adhere to the rules.

A simple example:
```
elf$ cat noel.c
#include <stdio.h>

int
main(int árgç, char **àrgv)
{
        char nöel[] = "Joyeux Nöel!";
        printf("%s\n", nöel);
        return 1;
}
elf$ wc -cwl noel.c
       9      18     127 noel.c
elf$ wc -mwl noel.c
       9      18     121 noel.c
elf$ cc noel.c
elf$ ./a.out
Joyeux Nöel!
elf$ od -c noel.c
0000000    #   i   n   c   l   u   d   e       <   s   t   d   i   o   .
0000020    h   >  \n  \n   i   n   t  \n   m   a   i   n   (   i   n   t
0000040      303 241   r   g 303 247   ,       c   h   a   r       *   *
0000060  303 240   r   g   v   )  \n   {  \n  \t   c   h   a   r       n
0000100  303 266   e   l   [   ]       =       "   J   o   y   u   e   x
0000120        N 303 266   e   l   !   "   ;  \n  \t   p   r   i   n   t
0000140    f   (   "   %   s   \   n   "   ,       n 303 266   e   l   )
0000160    ;  \n  \t   r   e   t   u   r   n       1   ;  \n   }  \n
0000177
elf$
```
So this is less radical than the idea of adding to Rule 2b the clause: `All identifiers count at most two (2) octets towards Rule 2b.`.  (Ixnay for now.)  

The switch of wording from bytes/octets to characters still retains the current spirit of Rule 2 (and also still printable on t-shirts or as tattoos) and remains C standard compliant 😁

**Reply by xexyl** [reply]
> To restate a point I made above that might having been overlooked:
> 
> > Or maybe the change to make is to count characters instead of octets.
> 
> The above minor change would be the simplest evolution of Rule 2 without drastically changing the rules as proposed by OP. So ...
> 
> Reading a C source file from standard input, apply the IOCCC source size rules as explained in the Guidelines. The source code is passed through on standard output. The source's net _**character**_ length is written to standard error; with -v option the net _**character**_ length, gross _**character**_ length, and matched keyword counts are written.
> 
> The entry's gross size in ~bytes~ _**characters**_ must be less than equal to 4993 ~bytes~ _**characters**_ in length.
> 
> The entry's net size in ~bytes~ _**characters**_ must be less than equal to 2503 ~bytes~ _**characters**_. The net size is computed as follows:
> 
> The size tool counts most C reserved words (keyword, secondary, and selected preprocessor keywords) as 1. The size tool counts all other ~octets~ _**characters**_ as 1 excluding ASCII whitespace, and excluding any ‘;’, ‘{’ or ‘}’ followed by ASCII whitespace, and excluding any ‘;’, ‘{’ or ‘}’ ~octet~ _**character**_ immediately before the end of file.
> 
> The should allow for better UTF8 multibyte support, which the C Standard has supported for some time now. This means that non-english speakers could present entries where function and variable names are no longer ASCII characters, but still adhere to the rules.
> 
> A simple example:
> 
> ```
> elf$ cat noel.c
> #include <stdio.h>
> 
> int
> main(int árgç, char **àrgv)
> {
>         char nöel[] = "Joyeux Nöel!";
>         printf("%s\n", nöel);
>         return 1;
> }
> elf$ wc -cwl noel.c
>        9      18     127 noel.c
> elf$ wc -mwl noel.c
>        9      18     121 noel.c
> elf$ cc noel.c
> elf$ ./a.out
> Joyeux Nöel!
> elf$ od -c noel.c
> 0000000    #   i   n   c   l   u   d   e       <   s   t   d   i   o   .
> 0000020    h   >  \n  \n   i   n   t  \n   m   a   i   n   (   i   n   t
> 0000040      303 241   r   g 303 247   ,       c   h   a   r       *   *
> 0000060  303 240   r   g   v   )  \n   {  \n  \t   c   h   a   r       n
> 0000100  303 266   e   l   [   ]       =       "   J   o   y   u   e   x
> 0000120        N 303 266   e   l   !   "   ;  \n  \t   p   r   i   n   t
> 0000140    f   (   "   %   s   \   n   "   ,       n 303 266   e   l   )
> 0000160    ;  \n  \t   r   e   t   u   r   n       1   ;  \n   }  \n
> 0000177
> elf$
> ```
> 

I did not know that! Interesting ... though I guess not all compilers do accept it. Still if clang supports it that's a good sign.

> So this is less radical than the idea of adding to Rule 2b the clause: `All identifiers count at most two (2) octets towards Rule 2b.`. (Ixnay for now.)
> 
> The switch of wording from bytes/octets to characters still retains the current spirit of Rule 2 (and also still printable on t-shirts or as tattoos) and remains C standard compliant 😁

Well as long as say, `a` still counts as the same as before I don't see the problem. Unless of course it makes it less likely that submissions will win if they have single character letters. All the more especially since many people, I am sure, have already written code for the next contest. Yusuke told us that he had a directory for the next contest and I am sure he already worked on some: though of course I don't want to speak for him. Still my only issue would be if it would result in less likely to win. The words you used at first, below, are what concerned me:

> Does compacting a program to use single letters for identifiers really help with obfuscation any more or is it a tired "trope"?

It certainly does help and it has other benefits too. But of course even if it doesn't help (per se - but that would depend on what the identifiers are) it's not the only part of the obfuscation so it shouldn't matter anyway. Still that question is what concerned me. It reminded me of the words along the lines of 'what was novel one year might be old the next'.

**Reply by xexyl** [reply]
> > To restate a point I made above that might having been overlooked:
> > > Or maybe the change to make is to count characters instead of octets.
> > 
> > 

> [...]

Of course I do not know the internals of `iocccsize` so it might be just fine: consider it a comment partly from ignorance but also concern, esp the last part I quoted.

**Reply by xexyl** [reply]
>  Yusuke told us that he had a directory for the next contest and I am sure he already worked on some: though of course I don't want to speak for him. 

By this I meant that I know he was not wanting the mock contest as it would only delay the next one. He told me during the review period of 2020 that he had several ideas in a file for future contests and one of the ones that he always felt was a 'must have' was the Enigma machine and he thanked me for doing it (a big honour from him). What code he has or has not written I of course do not know but since it has been over four years it's quite possible people have written code or at least started to. 

Anyway those were my concerns and maybe they're entirely irrelevant too. I hope so but I felt I should bring them up anyway just in case.

**Reply by SirWumpus** [reply]
> the internals of iocccsize so it might be just fine

After making this suggestion this morning about "octets -> characters", I went had a jiggery pokery at my `iocccsize` and it didn't take long to switch to handle UTF8 multibyte characters.  So thoughts on that are...

* should the current numbers for -v0 (default) and -v1 expressed in bytes be updated to characters?  OR
* should the output present the current values `2b-bytes 2a-bytes keyswords 2b-chars 2a-chars`?  OR
* should we go the way of `wc(1)` multibyte support and introduce `-m` (multibyte) option (since historically `-c` was octets (not characters go figure))?  So `iocccsize` would default to bytes and with `-m` switch to characters.

The 1st point is probably best choice, while the 2nd shows everything which can be nice for comparative debugging with `wc(1)` and editors.  The 3rd choice could result in users not paying attention when submitting (or make characters default output and have `-b` for byte counts).

All of the above is predicated on the idea that The Judges (why do I think of Judge Dredd each time) approve of the "octets -> characters" change.

**Reply by xexyl** [reply]
> > the internals of iocccsize so it might be just fine
> 
> After making this suggestion this morning about "octets -> characters", I went had a jiggery pokery at my `iocccsize` and it didn't take long to switch to handle UTF8 multibyte characters. So thoughts on that are...
> 
> * should the current numbers for -v0 (default) and -v1 expressed in bytes be updated to characters?  OR
> * should the output present the current values `2b-bytes 2a-bytes keyswords 2b-chars 2a-chars`?  OR
> * should we go the way of `wc(1)` multibyte support and introduce `-m` (multibyte) option (since historically `-c` was octets (not characters go figure))?  So `iocccsize` would default to bytes and with `-m` switch to characters.
> 
> The 1st point is probably best choice, while the 2nd shows everything which can be nice for comparative debugging with `wc(1)` and editors. The 3rd choice could result in users not paying attention when submitting (or make characters default output and have `-b` for byte counts).
> 
> All of the above is predicated on the idea that The Judges (why do I think of Judge Dredd each time) approve of the "octets -> characters" change.

Whilst I can't make the 'final decision' I can say that mkiocccentry actually warns about rule 2a and rule 2b violations so the only time they would have to worry about it is ahead of time when working on it or if they want to verify it won't break either rule before trying to submit.

But of course doing that is a good idea and helps one know what they have to do to decrease the size (whether adding new obfuscation - as I and I'm sure many others have done - or changing things - as I and I'm sure others have done - or maybe in some cases removing things - which I'm not sure if I have done) prior to packaging.

From my perspective then, as I do indeed do that, if this change is desired, it would be good to not have to make the submitters worry about how to invoke it (though in my past entries I had two Makefiles: a Makefile that builds and tests it and a Makefile that is submitted and the former had a rule/target that runs iocccsize on it - so it wouldn't be a big deal).

**Reply by xexyl** [reply]
> > > the internals of iocccsize so it might be just fine
> > 
> > 
> > After making this suggestion this morning about "octets -> characters", I went had a jiggery pokery at my `iocccsize` and it didn't take long to switch to handle UTF8 multibyte characters. So thoughts on that are...
> > 
> > * should the current numbers for -v0 (default) and -v1 expressed in bytes be updated to characters?  OR
> > * should the output present the current values `2b-bytes 2a-bytes keyswords 2b-chars 2a-chars`?  OR
> > * should we go the way of `wc(1)` multibyte support and introduce `-m` (multibyte) option (since historically `-c` was octets (not characters go figure))?  So `iocccsize` would default to bytes and with `-m` switch to characters.
> > 
> > The 1st point is probably best choice, while the 2nd shows everything which can be nice for comparative debugging with `wc(1)` and editors. The 3rd choice could result in users not paying attention when submitting (or make characters default output and have `-b` for byte counts).
> > All of the above is predicated on the idea that The Judges (why do I think of Judge Dredd each time) approve of the "octets -> characters" change.
> 
> Whilst I can't make the 'final decision' I can say that mkiocccentry actually warns about rule 2a and rule 2b violations so the only time they would have to worry about it is ahead of time when working on it or if they want to verify it won't break either rule before trying to submit.
> 
> But of course doing that is a good idea and helps one know what they have to do to decrease the size (whether adding new obfuscation - as I and I'm sure many others have done - or changing things - as I and I'm sure others have done - or maybe in some cases removing things - which I'm not sure if I have done) prior to packaging.
> 
> From my perspective then, as I do indeed do that, if this change is desired, it would be good to not have to make the submitters worry about how to invoke it (though in my past entries I had two Makefiles: a Makefile that builds and tests it and a Makefile that is submitted and the former had a rule/target that runs iocccsize on it - so it wouldn't be a big deal).

.. of course the guidelines do talk about the invocation so knowing how it should be used would be helpful to me - especially as I have some upcoming technical changes to fix. But that doesn't have to be done in the same commit of course.

**Reply by xexyl** [reply]
I will add something though: I do like the idea of having diacritics in code. That could be fun and it would be a new way of obfuscating, to be sure. It would also solve a problem with running out of letters as one could use for instance é, ë and so on, assuming it's all valid.

**Reply by xexyl** [reply]
> I will add something though: I do like the idea of having diacritics in code. That could be fun and it would be a new way of obfuscating, to be sure. It would also solve a problem with running out of letters as one could use for instance é, ë and so on, assuming it's all valid.

In fact it would have been very fun to do for the Enigma machine of 2020! I would have probably enjoyed that.

**Reply by SirWumpus** [reply]
> ... one could use for instance é, ë and so on, assuming it's all valid.

For the purpose of C I would think each character with/without diacritics would be treated differently, much like upper and lower case is different.  In a French dictionary they group the words by e, ë, é, è (not as just 'e').  Same would be true of Hirigana and Katakana, same alphabet, but different glyphs (katakana is used for foreign words typically).

**Reply by xexyl** [reply]
> > ... one could use for instance é, ë and so on, assuming it's all valid.
> 
> For the purpose of C I would think each character with/without diacritics would be treated differently, much like upper and lower case is different. In a French dictionary they group the words by e, ë, é, è (not as just 'e'). Same would be true of Hirigana and Katakana, same alphabet, but different glyphs (katakana is used for foreign words typically).

And then there's the debate in German: are they part of the alphabet or not. And then there's the Eszett (ẞ).

I am not fluent in German (far from it!) but I know a bit and I find it a beautiful language (I know many people find it the opposite though); French I know absolutely nothing about except I don't know that I like it. I don't know if that's true and I know there are some really funny things about it (something to do with numbers to not get into rude things) but maybe I just don't like the sound of it. Maybe I haven't heard it enough to let it grow on me. Either way I know next to nothing if not nothing about it.

I did have a thought that (and this is just a thought that popped into my head - I have absolutely no idea whether it's a good idea or not) maybe characters with diacritics (or the Eszett) might be counted as 2 bytes (or N+1 where N is what we had valid in the contest before) but I don't know what I would think about it either. It just popped into my head. If the size limit went back down I would definitely not want it but I guess that's not any more certain than your idea here.

I do **really appreciate** the dialogue btw and your clarifying things too.

**Reply by SirWumpus** [reply]
>  maybe characters with diacritics (or the Eszett) might be counted as 2 bytes

No point.  Once you start talking ***characters***, the idea of byte counting is pointless for text.  Anyway most european alphabets are UTF8 2 byte sequences, while the Asain ones I think are 3 and/or 4 byte sequences.  If you count by bytes, you penalise non-europeans for using their native alphabet.

**Reply by xexyl** [reply]
> > maybe characters with diacritics (or the Eszett) might be counted as 2 bytes
> 
> No point. Once you start talking _**characters**_, the idea of byte counting is pointless for text. Anyway most european alphabets are UTF8 2 byte sequences, while the Asain ones I think are 3 and/or 4 byte sequences. If you count by bytes, you penalise non-europeans for using their native alphabet.

That's true enough. It just popped into my head and thought I'd mention it. It would in a way penalise them but on the other hand they get to use extra characters. But then again ... that's not really nice and your point about size is also valid of course. Admittedly this was another concern of mine here anyway.

**Comment by SirWumpus** [comment]
@lcn2 Any thoughts on https://github.com/ioccc-src/mkiocccentry/discussions/903#discussioncomment-10019960 ?

**Reply by lcn2** [reply]
> @lcn2 Any thoughts on https://github.com/ioccc-src/mkiocccentry/discussions/903#discussioncomment-10019960 ?

None at this time that we wish to make.

Although is almost certain that such a shift would not occur without a lot of internal IOCCC judge discussion.  The changes to rule 2 sizes were controversial and complex enough (and are still not fully settled, BTW).  Our guess that anything close to what is being discussed might be post IOCCC28 at best (if it happens at all), but that is just a guess.

That's why we remain silent observers in this discussion.

**Reply by xexyl** [reply]
> > @lcn2 Any thoughts on [#903 (reply in thread)](https://github.com/ioccc-src/mkiocccentry/discussions/903#discussioncomment-10019960) ?
> 
> None at this time that we wish to make.
> 
> Although is almost certain that such a shift would not occur without a lot of internal IOCCC judge discussion. The changes to rule 2 sizes were controversial and complex enough (and are still not fully settled, BTW). Our guess that anything close to what is being discussed might be post IOCCC28 at best (if it happens at all), but that is just a guess.
> 
> That's why we remain silent observers in this discussion.

No longer silent though! :-)

It makes sense what you say though.

**Comment by SirWumpus** [comment]
> The changes to rule 2 sizes were controversial and complex enough (and are still not fully settled, BTW). 

That would be an interesting discussion to hear.

Though I see no conflict or issue between the size changes and UTF8 support.  One concerns size, the other embraces multibyte characters like the C standard.  The more I think about the idea to move from octets to characters, the more I believe its time to make an effort to move from 1980's thinking in terms of ASCII only.  UTF8 multibyte support has been around a long time now.

But that is just me.

**Reply by xexyl** [reply]
> > The changes to rule 2 sizes were controversial and complex enough (and are still not fully settled, BTW).
> 
> That would be an interesting discussion to hear.
> 
> Though I see no conflict or issue between the size changes and UTF8 support. One concerns size, the other embraces multibyte characters like the C standard. The more I think about the idea to move from octets to characters, the more I believe its time to make an effort to move from 1980's thinking in terms of ASCII only. UTF8 multibyte support has been around a long time now.
> 
> But that is just me.

I'd like to hear that discussion too but I guess it's something only judges get to hear. The other part is a nice idea and I'm starting to like it too .. though I guess not sure entirely.

Have a great rest of your day and thanks for the dialogue! Off for the night.

**Comment by SirWumpus** [comment]
> UTF-8 demands equal rights and respect as real characters! The dominance of slim ASCII must end! We are the UTF-8 People's Front for Freedom, Equality, and Brotherhood (oh that's been used already) - We are the UTF-8 Countless Tourist Defence Farce for One Voice, One World, One Character Set.

For comparison and future consideration I have implemented an extension to (the original) [iocccsize](https://github.com/SirWumpus/iocccsize/tree/chsize) with UTF-8 character counting.  This should allow to compare byte vs character size differences, with a goal to eventually change the Rule 2 wording to count by character, instead of bytes.

The `-v1` output has been enhanced...
```
$ iocccsize -v1 prog.c
2a_net 2b_gross 2a_mb_net 2b_mb_gross keywords
```

**Comment by lcn2** [comment]
After an extended conversation with Leo, we have determined that we will **NOT** change Rule 2a from bytes to UTF-8 characters.  We know that this puts UTF-8 variables and function names at a disadvantage, but so is ASCII variables and function names.  Leo was unhappy with the recent size change and was definite about not modifying the Rule 2a anytime soon. 

There is pondering about changing Rule 2b to consider counting UTF-8 characters, however, but that is an on-going discussion.  We might look into this for **IOCCC30**, however such a Rule 2b change would still have to fit within the Rule 2a byte limits.

Sorry (tm Canada 🇨🇦), this may not have been the answer you have wanted.

**Reply by SirWumpus** [reply]
This is fine.  You the Judges had made that statement before.  I did not think 29 or 30 would adopt this change.  I simply made the tweak, as diagnostic information, for far future consideration and comparative.  At least if the tool has the ability, you could look at the numbers in IOCCC29, IOCCC30, IOCCC31.  I'm thinking this is a future direction AFTER, the recent rule change runs its course.  Odds are my little UTF-8 usage example will remain an oddity.
